daemon/cmd/fqdn.go

```diff
@@ -61,97 +61,61 @@ const (
+// updateSelectors propagates the mapping of FQDNSelector to IP
+// to the policy engine.
+// First, it updates any selectors in the SelectorCache, then it triggers
+// all endpoints to push incremental updates via UpdatePolicyMaps.
+//
+// returns a WaitGroup that is done when all policymaps have been updated
+// and all endpoints referencing the IPs are able to pass traffic.
+func (d *Daemon) updateSelectors(ctx context.Context, selectors map[policyApi.FQDNSelector][]netip.Addr, ipcacheDoneChan <-chan struct{}) (wg *sync.WaitGroup) {
 	// There may be nothing to update - in this case, we exit and do not need
 	// to trigger policy updates for all endpoints.
-	if len(selectors) == 0 && len(selectorsWithoutIPs) == 0 {
+	if len(selectors) == 0 {
 		return &sync.WaitGroup{}
 	}
+	logger := log.WithField("qname", ctx.Value(requestNameKey{}))
 
+	// notifyWg is a waitgroup that is incremented for every "user" of a selector; i.e.
+	// every single SelectorPolicy. Once that selector has pushed out incremental changes
+	// to every relevant endpoint, the WaitGroup will be done.
 	notifyWg := &sync.WaitGroup{}
+	updateResult := policy.UpdateResultUnchanged
 	// Update mapping of selector to set of IPs in selector cache.
-	for selector, identitySlice := range selectors {
-		log.WithFields(logrus.Fields{
+	for selector, ips := range selectors {
+		logger.WithFields(logrus.Fields{
 			"fqdnSelectorString": selector,
-			"identitySlice":      identitySlice}).Debug("updating FQDN selector")
-		numIds := make([]identity.NumericIdentity, 0, len(identitySlice))
-		for _, numId := range identitySlice {
-			// Nil check here? Hopefully not necessary...
-			numIds = append(numIds, numId.ID)
-		}
-		d.policy.GetSelectorCache().UpdateFQDNSelector(selector, numIds, notifyWg)
+			"ips":                ips}).Debug("updating FQDN selector")
+		res := d.policy.GetSelectorCache().UpdateFQDNSelector(selector, ips, notifyWg)
+		updateResult |= res
 	}
 
-	if len(selectorsWithoutIPs) > 0 {
-		// Selectors which no longer map to IPs (due to TTL expiry, cache being
-		// cleared forcibly via CLI, etc.) still exist in the selector cache
-		// since policy is imported which allows it, but the selector does
-		// not map to any IPs anymore.
-		log.WithFields(logrus.Fields{
-			"fqdnSelectors": selectorsWithoutIPs,
-		}).Debug("removing all identities from FQDN selectors")
-		d.policy.GetSelectorCache().RemoveIdentitiesFQDNSelectors(selectorsWithoutIPs, notifyWg)
+	// UpdatePolicyMaps consumes notifyWG, and returns its own WaitGroup
+	// that is Done() when all endpoints have pushed their incremental changes
+	// down in to their bpf PolicyMap.
+	if updateResult&policy.UpdateResultUpdatePolicyMaps > 0 {
+		logger.Debug("FQDN selector update requires UpdatePolicyMaps.")
+		wg = d.endpointManager.UpdatePolicyMaps(ctx, notifyWg)
+	} else {
```

Comment:
Correct, if we removed `UpdatePolicyMaps` here we would break case (2). To be clear, I was not suggesting that we do that in this PR. If we introduce a mechanism to trigger the ipcache to call `UpdatePolicyMaps` without having a new IP, then technically we could remove this extra call to `UpdatePolicyMaps`. Functionally there's no impact when moving this call to `UpdatePolicyMaps` from here to the ipcache. But what it does do is centralize the logic for triggering the update of policymaps into the ipcache, so that other subsystems don't need to do it themselves.

That's what I meant by:

> For posterity, the idea discussed was to plumb extra metadata down to the ipcache so that it can call UpdatePolicyMaps to update endpoints that a selector now has a new identity relevant to them. Today, ipcache no-ops in this case because it relies on the FQDN selectorcache layer. 

If we feel like there's not much benefit to that, I'm open to that viewpoint. In general, I am interested in exploring the idea to move as much ipcache-related logic into ipcache, rather than potentially complicating call siteswith those details. We have been trending in the direction of ipcache becoming the ipcache is the central point for endpoints/policies, so I see it as a natural extension of that direction.

Reply:
:+1: thanks for clarifying. I agree.

(incidentally, this case (2) would make for an interesting lifecycle test... how does this new approach handle updates to ipcache? If multiple endpoints make DNS lookups for the same name, that will result in a single ipcache entry update right? Will that entry remain live until TTL expires and both endpoints stop using that name? Failing to keep the ipcache entry alive for the full zombie period could cause packet loss. In other words, the `RemoveMetadata()` call should only occur for the same ipcache resourceID once all namemanager usage of the corresponding IP has been garbage collected.)

Reply:
Good question. From my reading of the code, the `RemoveMetadata` is called within `maybeRemoveMetadataLocked()` of the NameManager during the GC routine.

The following steps occur in the GC logic (`(n *NameManager) GC`):

1. Collect list of `maybeStaleIPs` which is the entire dump of the DNS cache by IP
2. Mark all entries under `ep.DNSHistory` for each endpoint
3. Mark all entries under `ep.DNSZombies`
4. Collate list of all alive entries from steps (2) and (3) and replace the global DNS cache with the new set of only alive entries
5. Skip each entry from `maybeStaleIPs` if it exists in the new DNS cache, call `RemoveMetadata` on the ones that don't exist in the DNS cache

From the above high level steps, I don't believe there's a possibility to prematurely delete an entry that would result in a deleted ipcache entry while the DNS entry is still a zombie.

Reply:
(Answering your general point, I do think lifecycle tests like this would greatly beneficial to have in the codebase if we don't already.)

Reply:
The policy engine has two sources of incremental updates (IPCache and NameManager), so it has two separate callers of UpdatePolicyMaps. In fact, the ipcache needs to call UpdatePolicyMaps twice, since it needs a strict (add, commit, delete) ordering.

Chris is right that, in the event cases 1 and 2 are *both* relevant to an FQDN update, we will waste some time doing 3 calls to `UpdatePolicyMaps()` when only 2 are required. However, the FQDN and ipcache flows are sufficiently distinct from one another, so it's not easy to coalesce these calls down. Especially because all 3 callers require the update to be functionally synchronous.

One easy change would be to put UpdatePolicyMaps behind a controller / trigger, and coalesce pending calls. Otherwise, we need to start thinking about a more streamlined design for incremental updates.

However, I'd like to understand if this is has a real-world performance implication before spending a lot of time thinking about this. My intuition is that, most of the time, only an ipcache or fqdn incremental update will be relevant. The case where a FQDN response both allocates a new IP *and* newly-selects an existing should be fairly infrequent.

Reply:
Now that I think about this a bit more, and reminding myself of two critical facts:
1. we don't regenerate endpoints until k8s caches have synced
2. un-regenerated pods reference an **old ipcache map** on restart

I actually believe this PR is **not a regression**. Namely, in `main`, if a new IP is allocated, we will add it to the endpoint's PolicyMap, but the endpoint's ipcache **won't have that IP**. I need to do a quick experiment, but this might turn out to be nothing.

Reply:
I should have tested this days ago. Apologies for the much ado about nothing.

Conclusion: In a caches-unsynched agent, we currently reject all DNS queries (#29508, bug). When that bug is fixed, we *still can't plumb new responses*, because there are no selectors, because endpoints haven't regenerated yet. Because there are no selectors, there is no day in propagating selector changes.

So this PR neither regresses on behavior nor adds an additional delay. This conversation can be resolved, and I learned a valuable lesson: test! before! worrying!